File: src/main/java/com/procurement/submission/model/ocds/Address.java

package com.procurement.submission.model.ocds;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import lombok.Getter;
import org.apache.commons.lang.builder.EqualsBuilder;
import org.apache.commons.lang.builder.HashCodeBuilder;

@Getter
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonPropertyOrder({
    "streetAddress",
    "locality",
    "region",
    "postalCode",
    "countryName"
})
public class Address {

    @NotNull
    @JsonProperty("streetAddress")
    private final String streetAddress;

    @NotNull
    @JsonProperty("locality")
    private final String locality;

    @NotNull
    @JsonProperty("region")
    private final String region;

    @JsonProperty("postalCode")
    private final String postalCode;

    @NotNull
    @JsonProperty("countryName")
    private final String countryName;

    @JsonCreator
    public Address(@JsonProperty("streetAddress") final String streetAddress,
                   @JsonProperty("locality") final String locality,
                   @JsonProperty("region") final String region,
                   @JsonProperty("postalCode") final String postalCode,
                   @JsonProperty("countryName") final String countryName) {
        this.streetAddress = streetAddress;
        this.locality = locality;
        this.region = region;
        this.postalCode = postalCode;
        this.countryName = countryName;
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder().append(streetAddress)
                                    .append(locality)
                                    .append(region)
                                    .append(postalCode)
                                    .append(countryName)
                                    .toHashCode();
    }

    @Override
    public boolean equals(final Object other) {
        if (other == this) {
            return true;
        }
        if (!(other instanceof Address)) {
            return false;
        }
        final Address rhs = (Address) other;
        return new EqualsBuilder().append(streetAddress, rhs.streetAddress)
                                  .append(locality, rhs.locality)
                                  .append(region, rhs.region)
                                  .append(postalCode, rhs.postalCode)
                                  .append(countryName, rhs.countryName)
                                  .isEquals();
    }
}

File: src/main/kotlin/com/procurement/submission/model/dto/ocds/Address.kt
package com.procurement.access.model.dto.ocds

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonInclude

@JsonInclude(JsonInclude.Include.NON_NULL)
data class Address @JsonCreator constructor(

        val streetAddress: String,

        val postalCode: String?,

        val addressDetails: AddressDetails
)

@JsonInclude(JsonInclude.Include.NON_NULL)
data class AddressDetails(

        val country: CountryDetails,

        val region: RegionDetails,

        val locality: LocalityDetails
)

@JsonInclude(JsonInclude.Include.NON_NULL)
data class CountryDetails(

        var scheme: String?,

        val id: String,

        var description: String?,

        var uri: String?
)

@JsonInclude(JsonInclude.Include.NON_NULL)
data class RegionDetails(

        var scheme: String?,

        val id: String,

        var description: String?,

        var uri: String?
)

@JsonInclude(JsonInclude.Include.NON_NULL)
data class LocalityDetails(

        var scheme: String,

        val id: String,

        var description: String,

        var uri: String?
)